### PR TITLE
#43 google auth karleen

### LINKED ISSUE #43: Task1: Karleen- Authenticate with Google
Feature: User Authentication and Login with Google

Scenario: Unauthenticated user clicks sign-in button and fails Google sign-in
    Given I am an unauthenticated user
    When I click the sign-in button
    And I am unable to successfully sign in or create an account with Google
    Then I see the homepage

Scenario: Unauthenticated user clicks sign-in button and successfully logs in with Google
    Given I am an unauthenticated user
    When I click the sign-in button
    And I am able to successfully sign in or create an account with Google
    Then I see the dashboard

Scenario: Authenticated user returns to the webpage and remains logged in
    Given I am an authenticated user
    When I leave the webpage
    And I come back to the webpage
    Then I am still logged in

Technical Task:
Authenticate with Google
Receive the token 
Save the token in the browser
Configure redirect url to dashboard

### PR DESCRIPTION
**Pull request by karleenmsrichards**
## Description

I have implemented google auth in the front end and saved the JWT token to local storage.

Fixes #43 

## Checklist:

- My code follows the style guidelines of this project
- ✅  I have carefully reviewed my own code
- ✅ I have commented my code
- I have updated any documentation

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
-		<Route path="/" element={<Home />} />
-		<Route path="/contact-us" element={<ContactUs />} />
-	</Routes>
+		<Header />
+		<Routes>
+			<Route path="/" element={<Home />} />
+			<Route path="/dashboard" element={<Dashboard />} />
+			<Route path="/contact-us" element={<ContactUs />} />
+		</Routes>
+		<Footer />
```
Comment (jarrodruby) [inline client/src/App.js:17]: do we want this much indentation?

File: client/src/customHooks/useAuth.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    });
+    google.accounts.id.prompt();
+  }
+
+  function handleCallbackResponse(response) {
+    if (response && response.credential) {
+      console.log("Encoded JWT ID token: " + response.credential);
+      let userObject = jwtDecode(response.credential);
+      console.log(userObject);
+      // Save the token in localStorage
```
Comment (jarrodruby) [inline client/src/customHooks/useAuth.js:29]: we don't need this comment as it is obvious. See this concept https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/

File: server/api.js [hunk 3]
```diff
@@ -9,4 +10,17 @@ router.get("/", (_, res) => {
 	res.json({ message: "Hello, world!" });
 });
 
+router.get("/clientId", (req, res) => {
+	try {
+	const clientId = process.env.REACT_APP_CLIENT_ID;
+	if (!clientId) {
+		throw new Error("Client ID not found.");
+	}
+res.json({ clientId }).status(200);
```
Comment (jarrodruby) [inline server/api.js:19]: indentation

File: client/src/Components/Header.jsx [hunk 4]
```diff
@@ -1,15 +1,36 @@
...
+            <Button variant="outlined" onClick={handleSignUp}>
+              Sign Up
+            </Button>
+            <Button variant="outlined" onClick={handleSignUp}>
+              Sign In
+            </Button>
+          </Box>
+        ) : (
+          <></>
+        )}
```
Comment (jarrodruby) [inline client/src/Components/Header.jsx:24]: Can we try this? Not sure if that will work but would be consistent with the approach below. For the team to discuss
`
{ !user && (
          <Box>
            <Button variant="outlined" onClick={handleSignUp}>
              Sign Up
            </Button>
            <Button variant="outlined" onClick={handleSignUp}>
              Sign In
            </Button>
          </Box>
        ) 
`

File: client/src/Components/Header.jsx [hunk 5]
```diff
@@ -1,15 +1,36 @@
...
+      <Box sx={{ display: "flex", justifyContent: "flex-end", py: 2, px: 5, gap: 5 }}>
+        { !user ? (
+          <Box>
+            <Button variant="outlined" onClick={handleSignUp}>
+              Sign Up
+            </Button>
+            <Button variant="outlined" onClick={handleSignUp}>
+              Sign In
+            </Button>
+          </Box>
```
Comment (jarrodruby) [inline client/src/Components/Header.jsx:21]: redundant code
`
            <Button variant="outlined" onClick={handleSignUp}>
              { user ?  'Sign Up' : 'Sign Up'}
            </Button>
`

File: client/src/customHooks/useAuth.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        console.error("Error fetching client ID:", error);
+      }
+    }
+    fetchClientId();
+  }, []);
+
+  useEffect(() => {
+    // Check if the user is already authenticated by looking for the token in localStorage
+    const token = localStorage.getItem("jwtToken");
+    console.log("local");
```
Comment (jarrodruby) [inline client/src/customHooks/useAuth.js:54]: do we need this log?

File: client/src/customHooks/useAuth.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    if (token) {
```
Comment (jarrodruby) [inline client/src/customHooks/useAuth.js:55]: javascript equality. Do we want to use this approach? Up to the team. FYI https://dorey.github.io/JavaScript-Equality-Table/

File: server/api.js [hunk 8]
```diff
@@ -9,4 +10,17 @@ router.get("/", (_, res) => {
...
+	} catch (error) {
+	logger.error("Error fetching clientId:", error.message);
+	res.status(500).json({ error: "Internal server error" });
```
Comment (jarrodruby) [inline server/api.js:22]: Error messages are good. We don't eant the user to know too much

File: client/src/customHooks/useAuth.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
+import { useState, useEffect } from "react";
+import jwtDecode from "jwt-decode";
+import axios from "axios";
+import { useNavigate } from "react-router-dom";
+
+const useAuth = () => {
+  const navigate = useNavigate();
+  const [user, setUser] = useState(null);
+  const [clientId, setClientId] = useState("");
```
Comment (jarrodruby) [inline client/src/customHooks/useAuth.js:9]: `useState(null)` would be consistent unless we have a good reason. Empty strings in js are evil

**Review by jarrodruby — APPROVED** [review]

File: client/src/customHooks/useAuth.js (hunk 2 above)
Comment (karleenmsrichards) [reply client/src/customHooks/useAuth.js:29]: Fixed

File: client/src/customHooks/useAuth.js (hunk 9 above)
Comment (karleenmsrichards) [reply client/src/customHooks/useAuth.js:9]: Fixed.

File: client/src/customHooks/useAuth.js (hunk 6 above)
Comment (karleenmsrichards) [reply client/src/customHooks/useAuth.js:54]: No, it was for testing. Removed.

File: client/src/App.js (hunk 1 above)
Comment (karleenmsrichards) [reply client/src/App.js:17]: Manually adjusted  indentation until I can fix eslint.

Comment (karleenmsrichards) [reply client/src/App.js:17]: Need to fix eslint.

File: client/src/customHooks/useAuth.js (hunk 7 above)
Comment (karleenmsrichards) [reply client/src/customHooks/useAuth.js:55]: do you mean I should use ```if (token === true)```

File: client/src/Components/Header.jsx (hunk 5 above)
Comment (karleenmsrichards) [reply client/src/Components/Header.jsx:21]: Resolved

**Pull request merged**